From 85b5d17fb40ecb192d33832fb1855b70a8b0e99c Mon Sep 17 00:00:00 2001
From: Liming Wang <liming.wang@windriver.com>
Date: Mon, 21 Jan 2013 16:11:21 +0800
Subject: [PATCH 1/2] amba-pl011, kgdboc: add in_kdb_printk to avoid calltrace
 in preempt-rt

Add check function in_kdb_printk to avoid calling spin_lock with the
irq disabled in preemp-rt kernel if KGDBoC is triggered. Otherwise the
call path trips the rtmutex debug code:

BUG: sleeping function called from invalid context at kernel/rtmutex.c:646

This change is inspired by serial8250_console_write() function in
drivers/tty/serial/8250/8250.c file.

Signed-off-by: Liming Wang <liming.wang@windriver.com>
---
 drivers/tty/serial/amba-pl011.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/amba-pl011.c b/drivers/tty/serial/amba-pl011.c
index 3139b7d..41f045c 100644
--- a/drivers/tty/serial/amba-pl011.c
+++ b/drivers/tty/serial/amba-pl011.c
@@ -52,6 +52,7 @@
 #include <linux/scatterlist.h>
 #include <linux/delay.h>
 #include <linux/types.h>
+#include <linux/kdb.h>
 
 #include <asm/io.h>
 #include <asm/sizes.h>
@@ -1794,7 +1795,7 @@ pl011_console_write(struct console *co, const char *s, unsigned int count)
 	local_irq_save_nort(flags);
 	if (uap->port.sysrq)
 		locked = 0;
-	else if (oops_in_progress)
+	else if (oops_in_progress || in_kdb_printk())
 		locked = spin_trylock(&uap->port.lock);
 	else
 		spin_lock(&uap->port.lock);
-- 
1.7.9.7

